addons/hashes/fnc_encodeJSON.sqf

```diff
@@ -0,0 +1,88 @@
+#include "script_component.hpp"
+/* ----------------------------------------------------------------------------
+Function: CBA_fnc_encodeJSON
+
+Description:
+    Serializes input to a JSON string. Can handle
+    - ARRAY
+    - BOOL
+    - CONTROL
+    - GROUP
+    - LOCATION
+    - NAMESPACE
+    - NIL (ANY)
+    - NUMBER
+    - OBJECT
+    - STRING
+    - TASK
+    - TEAM_MEMBER
+
+Parameters:
+    _object - Object to serialize. <ARRAY, ...>
+
+Returns:
+    _json   - JSON string containing serialized object.
+
+Examples:
+    (begin example)
+        private _settings = call CBA_fnc_createNamespace;
+        _settings setVariable ["enabled", true];
+        private _json = [_settings] call CBA_fnc_encodeJSON;
+    (end)
+
+Author:
+    BaerMitUmlaut
+---------------------------------------------------------------------------- */
+SCRIPT(encodeJSON);
+params ["_object"];
+
+if (isNil "_object") exitWith { "null" };
+
+switch (typeName _object) do {
+    case "SCALAR";
+    case "BOOL": {
+        str _object;
+    };
+
+    case "STRING": {
+        {
+            _object = [_object, _x#0, _x#1] call CBA_fnc_replace;
+        } forEach [
+            ["""", "\"""],
+            ["\", "\\"],
+            [toString [8], "\b"],
+            [toString [12], "\f"],
+            [endl, "\n"],
+            [toString [13], "\r"],
+            [toString [9], "\t"]
+        ];
+        str _object
+    };
+
+    case "ARRAY": {
+        if ([_object] call CBA_fnc_isHash) then {
+            private _json = (([_object] call CBA_fnc_hashKeys) apply {
+                private _name = _x;
+                private _value = [_object, _name] call CBA_fnc_hashGet;
+
+                format ["%1: %2", [_name] call CBA_fnc_encodeJSON, [_value] call CBA_fnc_encodeJSON]
+            }) joinString ", ";
+            "{" + _json + "}"
+        } else {
+            private _json = (_object apply {[_x] call CBA_fnc_encodeJSON}) joinString ", ";
+            "[" + _json + "]"
+        };
+    };
+
+    default {
+        if (isNull _object) exitWith { "null" };
```

Comment:
I have this urge to turn objects like the `player` into JSON files.

There should be a check if the _object type supports the `isNull` command.
JSON has no `undefined`, apparently, but not supported types should be treated as `null` with a RPT log message (either ERROR_n or WARNING_n).

```sqf
player call CBA_fnc_encodeJSON; 
```

<details><summary>Error</summary>

```sqf
16:55:23 Error in expression <
"[" + _json + "]"
};
};

default {
if (isNull _object) exitWith { "null" };

pr>
16:55:23   Error position: <isNull _object) exitWith { "null" };

pr>
16:55:23   Error isnull: Type code, expected Object,Group,Script,Config entry,Display (dialog),Control,Network Object,Task,Location
16:55:23 File \x\cba\addons\hashes\fnc_encodeJSON.sqf [CBA_fnc_encodeJSON]..., line 1910
16:55:23 Error in expression <
"[" + _json + "]"
};
};

default {
if (isNull _object) exitWith { "null" };

pr>
16:55:23   Error position: <isNull _object) exitWith { "null" };

pr>
16:55:23   Error Generic error in expression
16:55:23 File \x\cba\addons\hashes\fnc_encodeJSON.sqf [CBA_fnc_encodeJSON]..., line 1910
```

</details>